### PR TITLE
Learn uid order by update time desc

### PR DESCRIPTION
**Pull request by mingwli**
[AB#400391](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/400391)

- use `isLearn` to determine order logic on duplicate-uid

### CODE REVIEW

File: src/docfx/build/xref/InternalXrefMapBuilder.cs [hunk 1]
```diff
@@ -137,12 +140,19 @@ private InternalXrefSpec[] AggregateXrefSpecs(string uid, InternalXrefSpec[] spe
...
 
-            return specsWithSameUid
+            var specs = specsWithSameUid
                    .OrderByDescending(spec => spec.Monikers.HasMonikers
                         ? spec.Monikers.Select(moniker => _monikerProvider.GetMonikerOrder(moniker)).Max()
-                        : int.MaxValue)
-                   .ThenBy(spec => spec.DeclaringFile)
-                   .ToArray();
+                        : int.MaxValue);
+            if (_config.IsLearn)
```
Comment (yufeih) [inline src/docfx/build/xref/InternalXrefMapBuilder.cs:147]: Wrong order?

File: src/docfx/build/xref/InternalXrefMapBuilder.cs [hunk 2]
```diff
@@ -137,12 +140,19 @@ private InternalXrefSpec[] AggregateXrefSpecs(string uid, InternalXrefSpec[] spe
...
+            {
+                specs.ThenBy(spec => spec.DeclaringFile);
+            }
+            else
+            {
+                specs.ThenByDescending(spec => _contributionProvider.GetCommitInfo(spec.DeclaringFile).updatedDateTime);
```
Comment (yufeih) [inline src/docfx/build/xref/InternalXrefMapBuilder.cs:153]: Use `RepositoryProvider.GetCommitHistory`.

File: docs/specs/learn.yml [hunk 3]
```diff
@@ -1484,3 +1478,85 @@ outputs:
...
+        uid: unit
+outputs:
+  path.mta.json:
+  path.raw.page.json:
+  b-module.mta.json:
+  b-module.raw.page.json:
+  b-unit.mta.json:
+  b-unit.raw.page.json:
+  full-dependent-list.txt:
+  .errors.log:
```
Comment (yufeih) [inline docs/specs/learn.yml:1538]: verify there is `duplicate-uid` error

File: docs/specs/learn.yml [hunk 4]
```diff
@@ -1484,3 +1478,85 @@ outputs:
         }
       ]
     }
+---
+# Pick uid by commit time latest when conflict
```
Comment (live1206) [inline docs/specs/learn.yml:1482]: specify for `learn`

File: src/docfx/build/xref/InternalXrefMapBuilder.cs (hunk 1 above)
Comment (live1206) [inline src/docfx/build/xref/InternalXrefMapBuilder.cs:147]: Can we add a TODO to clean this up later?

File: src/docfx/build/xref/InternalXrefMapBuilder.cs [hunk 5]
```diff
@@ -137,12 +143,29 @@ private InternalXrefSpec[] AggregateXrefSpecs(string uid, InternalXrefSpec[] spe
...
+        private DateTime GetUpdateTime(FilePath file)
+        {
+            var fullPath = _input.TryGetOriginalPhysicalPath(file);
+            if (fullPath is null)
+            {
+                return default;
+            }
+
+            var commits = _repositoryProvider.GetCommitHistory(fullPath.Value).commits;
+            var result = commits.FirstOrDefault()?.Time.UtcDateTime ?? default;
```
Comment (herohua) [inline src/docfx/build/xref/InternalXrefMapBuilder.cs:167]: >result [](http://example.com/codeflow?start=16&length=6)

simplify to `return commits.FirstOrDefault()?.Time.UtcDateTime ?? default`

File: src/docfx/build/xref/InternalXrefMapBuilder.cs [hunk 6]
```diff
@@ -137,12 +143,29 @@ private InternalXrefSpec[] AggregateXrefSpecs(string uid, InternalXrefSpec[] spe
...
                         ? spec.Monikers.Select(moniker => _monikerProvider.GetMonikerOrder(moniker)).Max()
-                        : int.MaxValue)
-                   .ThenBy(spec => spec.DeclaringFile)
-                   .ToArray();
+                        : int.MaxValue);
+
+            // TODO: clean up after new loc pipeline
+            specs = _config.IsLearn
+                ? specs.ThenByDescending(spec => GetUpdateTime(spec.DeclaringFile))
+                : specs.ThenBy(spec => spec.DeclaringFile);
```
Comment (herohua) [inline src/docfx/build/xref/InternalXrefMapBuilder.cs:154]: if two files are of the same latest update time, should we order by file name?

**Review by herohua — APPROVED** [review]
:shipit:

File: src/docfx/build/xref/InternalXrefMapBuilder.cs (hunk 6 above)
Comment (mingwli) [reply src/docfx/build/xref/InternalXrefMapBuilder.cs:154]: Good thought! Though that would make output stable, but considering the loc deletion case, random picking one could better serve that purpose.

File: test/docfx.Test/TestUtility.cs [hunk 7]
```diff
@@ -102,6 +102,7 @@ static void Throw(string stackTrace, string message, string detailMessage, strin
 
             foreach (var commit in commits.Reverse())
             {
+                // TODO: move outside
```
Comment (yufeih) [inline test/docfx.Test/TestUtility.cs:105]: What does this TODO mean?

**Review by yufeih — APPROVED** [review]

**Pull request merged**